core/trino-main/src/main/java/io/trino/server/security/oauth2/OAuth2Service.java

```diff
@@ -104,63 +157,79 @@ public OAuthChallenge startWebUiChallenge(URI callbackUri)
         else {
             nonce = Optional.empty();
         }
-        return new OAuthChallenge(client.getAuthorizationUri(state, callbackUri, nonce.map(OAuth2Service::hashNonce)), challengeExpiration, nonce);
-    }
-
-    public URI startRestChallenge(URI callbackUri, UUID authId)
-    {
-        String state = Jwts.builder()
-                .signWith(SignatureAlgorithm.HS256, stateHmac)
-                .setId(authId.toString())
-                .setAudience(STATE_AUDIENCE_REST)
-                .setExpiration(Date.from(Instant.now().plus(challengeTimeout)))
-                .compact();
-        return client.getAuthorizationUri(state, callbackUri, Optional.empty());
+        Response.ResponseBuilder response = Response.seeOther(
+                client.getAuthorizationUri(
+                        state,
+                        uriInfo.getBaseUri().resolve(CALLBACK_ENDPOINT),
+                        nonce.map(OAuth2Service::hashNonce)));
+        nonce.ifPresent(nce -> response.cookie(NonceCookie.create(nce, challengeExpiration)));
+        return response.build();
     }
 
-    public OAuthResult finishChallenge(Optional<UUID> authId, String code, URI callbackUri, Optional<String> nonce)
+    public Response finishOAuth2Challenge(String state, String code, URI callbackUri, Optional<String> nonce)
             throws ChallengeFailedException
     {
-        requireNonNull(callbackUri, "callbackUri is null");
-        requireNonNull(authId, "authId is null");
-        requireNonNull(code, "code is null");
-
-        // fetch access token
-        AccessToken accessToken = client.getAccessToken(code, callbackUri);
-
-        // validate access token is trusted by this server
-        Claims parsedToken = Jwts.parser()
-                .setSigningKeyResolver(signingKeyResolver)
-                .parseClaimsJws(accessToken.getAccessToken())
-                .getBody();
+        String subjectState;
+        Optional<OAuth2TokenHandler> handler;
+        try {
+            Claims stateClaims = parseState(state);
+            subjectState = stateClaims.get(SUBJECT_STATE_CLAIM, String.class);
+            handler = tokenHandler.filter(h -> h.handlesSubject(stateClaims.getSubject()));
```

Comment:
I think we can (and should) simplify this.  I believe we only have two cases: web UI auth and query client auth.  Only the latter will have an authId, so if that claim is present, we know the handle should be used.  With that we can drop the `handlesSubject` method.

Reply:
👍